tests/returntypes/test_jax_jit_qnode_new.py

```diff
@@ -0,0 +1,598 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Integration tests for using the jax interface and its jittable variant with
+a QNode"""
+
+import pytest
+
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane import qnode
+from pennylane.tape import QuantumTape
+
+jit_qubit_device_and_diff_method = [
+    ["default.qubit", "backprop", "forward"],
+    # Jit
+    ["default.qubit", "finite-diff", "backward"],
+    ["default.qubit", "parameter-shift", "backward"],
+    ["default.qubit", "adjoint", "forward"],
+    ["default.qubit", "adjoint", "backward"],
+]
+
+pytestmark = pytest.mark.jax
+
+jax = pytest.importorskip("jax")
+jnp = jax.numpy
+
+from jax.config import config
+
+config.update("jax_enable_x64", True)
+
+jacobian_fn = [jax.jacobian, jax.jacrev, jax.jacfwd]
+
+
+@pytest.mark.parametrize("dev_name,diff_method,mode", jit_qubit_device_and_diff_method)
+@pytest.mark.parametrize("jacobian", jacobian_fn)
+class TestJIT:
+    """Test JAX JIT integration with the QNode and automatic resolution of the
+    correct JAX interface variant."""
+
+    def test_gradient(self, dev_name, diff_method, mode, jacobian, tol):
+        """Test derivative calculation of a scalar valued QNode"""
+        dev = qml.device(dev_name, wires=1)
+
+        if diff_method == "adjoint":
+            pytest.xfail(reason="The adjoint method is not using host-callback currently")
+
+        @jax.jit
+        @qnode(dev, diff_method=diff_method, interface="jax-jit", mode=mode)
```

Comment:
I think the best practice for jitting is to apply it on the outer most function, in this case jax.jacobian.

Reply:
Sure, changed this.